File: assignment 4/c/programs/8.c
#include <stdio.h>
#include <malloc.h>

struct Node
{
    int data;
    struct Node *next;
};

struct Node *createNode(int data)
{
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

struct Node *insert(struct Node *head, int data)
{
    struct Node *newNode = createNode(data);

    if (head == NULL)
    {
        head = newNode;
    }
    else
    {
        struct Node *temp = head;
        while (temp->next != NULL)
        {
            temp = temp->next;
        }
        temp->next = newNode;
    }

    return head;
}

void printList(struct Node *head)
{
    while (head != NULL)
    {
        printf("%d ", head->data);
        head = head->next;
    }
    printf("\n");
}

struct Node *swapPairs(struct Node *head)
{
    if (head == NULL || head->next == NULL)
    {
        return head;
    }

    struct Node *newHead = head->next; // New head after swapping
    struct Node *prev = NULL;

    while (head != NULL && head->next != NULL)
    {
        struct Node *nextNode = head->next->next; // Save the next pair

        // Swap the current pair
        head->next->next = head;
        if (prev != NULL)
        {
            prev->next = head->next;
        }
        head->next = nextNode;

        // Move pointers forward
        prev = head;
        head = nextNode;
    }

    return newHead;
}

int main()
{
    struct Node *head = NULL;
    int num, data;

    printf("Enter the number of elements in the list: ");
    scanf("%d", &num);

    printf("Enter the elements:\n");
    for (int i = 0; i < num; i++)
    {
        scanf("%d", &data);
        head = insert(head, data);
    }

    printf("Original List: ");
    printList(head);

    head = swapPairs(head);

    printf("List after Pairwise Swapping: ");
    printList(head);

    return 0;
}

File: assignment 4/python/programs/8.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


def create_node(data):
    return Node(data)


def insert(head, data):
    new_node = create_node(data)

    if head is None:
        head = new_node
    else:
        temp = head
        while temp.next is not None:
            temp = temp.next
        temp.next = new_node

    return head


def print_list(head):
    while head is not None:
        print(head.data, end=" ")
        head = head.next
    print()


def swap_pairs(head):
    if head is None or head.next is None:
        return head

    new_head = head.next
    prev = None

    while head is not None and head.next is not None:
        next_node = head.next.next

        head.next.next = head
        if prev is not None:
            prev.next = head.next
        head.next = next_node

        prev = head
        head = next_node

    return new_head


if __name__ == "__main__":
    head = None

    num = int(input("Enter the number of elements in the list: "))
    print("Enter the elements:")
    for _ in range(num):
        data = int(input())
        head = insert(head, data)

    print("Original List:", end=" ")
    print_list(head)

    head = swap_pairs(head)

    print("List after Pairwise Swapping:", end=" ")
    print_list(head)
